tools/pubsys/src/aws/validate_ami/ami.rs

```diff
@@ -0,0 +1,148 @@
+use futures::future::{join, ready};
+use futures::stream::{FuturesUnordered, StreamExt};
+use log::{info, trace};
+use serde::{Deserialize, Serialize};
+use snafu::ResultExt;
+use std::collections::HashMap;
+
+/// Structure of the EC2 image fields that should be validated
+#[derive(Serialize, Deserialize, PartialEq, Eq, Hash, Debug, Clone)]
+#[serde(rename_all = "PascalCase")]
+pub(crate) struct ImageDef {
+    /// The id of the EC2 image
+    pub(crate) image_id: String,
+
+    /// Whether or not the EC2 image is public
+    pub(crate) public: bool,
+
+    /// Whether or not the EC2 image supports Elastic Network Adapter
+    pub(crate) ena_support: bool,
+
+    /// The level of the EC2 image's Single Root I/O Virtualization support
+    pub(crate) sriov_net_support: String,
+}
+
+impl From<Image> for ImageDef {
+    fn from(image: Image) -> Self {
+        Self {
+            image_id: image.image_id().unwrap_or_default().to_string(),
+            public: image.public().unwrap_or_default(),
+            ena_support: image.ena_support().unwrap_or_default(),
+            sriov_net_support: image.sriov_net_support().unwrap_or_default().to_string(),
+        }
+    }
+}
+
+impl ImageDef {
+    // Creates a new ImageDef with a given image_id and expected values for public, ena_support,
+    // and sriov_net_support
+    pub(crate) fn expected(image_id: String) -> Self {
+        Self {
+            image_id,
+            public: true,
+            ena_support: true,
+            sriov_net_support: "simple".to_string(),
+        }
+    }
+}
+
+pub(crate) async fn describe_images<'a>(
+    clients: &'a HashMap<Region, Ec2Client>,
+    image_ids: &HashMap<Region, Vec<String>>,
+) -> HashMap<&'a Region, Result<HashMap<String, ImageDef>>> {
+    // Build requests for images; we have to request with a regional client so we split them by
+    // region
+    let mut requests = Vec::with_capacity(clients.len());
+    for region in clients.keys() {
+        trace!("Requesting images in {}", region);
+        let ec2_client: &Ec2Client = &clients[region];
+        let get_future = describe_images_in_region(
+            region,
+            ec2_client,
+            image_ids
+                .get(region)
+                .map(|i| i.to_owned())
+                .unwrap_or(vec![]),
+        );
+
+        requests.push(join(ready(region), get_future));
+    }
+
+    // Send requests in parallel and wait for responses, collecting results into a list.
+    requests
+        .into_iter()
+        .collect::<FuturesUnordered<_>>()
+        .collect()
+        .await
+}
+
+/// Fetches all images in a single region
+pub(crate) async fn describe_images_in_region(
+    region: &Region,
+    client: &Ec2Client,
+    image_ids: Vec<String>,
+) -> Result<HashMap<String, ImageDef>> {
+    info!("Retrieving images in {}", region.to_string());
+    let mut images = HashMap::new();
+
+    // Send the request
+    let mut get_future = client
```

Comment:
This is applicable to the implementation in your previous PR as well. I can't quite tell from reading this if things are happening truly in parallel? Can you tell from your tests how the performance will scale to 30 regions with 100s of AMIs in each?

I think the key question is whether the `send()` function is returning a *unordered* collection of futures.

Reply:
The requests are sent completely in parallel. These image requests are fast so the difference isn't really noticeable. For the SSM parameters, before they were parallelized, the total duration was over 3 minutes. After parallelizing, it took 1 minute (which is how long it takes to get the parameters from the region with the highest latency, even if that region is the only one to be queried).

If there were 30 regions with 100s of AMIs, the runtime would be the same as for a single region with the highest latency of those 30.